maas/resource_maas_boot_source.go

```diff
@@ -1,12 +1,146 @@
+		UpdateContext: resourceBootSourceUpdate,
+		DeleteContext: resourceBootSourceDelete,
+
+		Schema: map[string]*schema.Schema{
+			"created": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The creation time of the boot source.",
+			},
+			"keyring_data": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "The data on the keyring for the boot source.",
+				ConflictsWith: []string{"keyring_filename"},
+			},
+			"keyring_filename": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "The filename on the keyring for the boot source.",
+				ConflictsWith: []string{"keyring_data"},
+			},
+			"updated": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The time of most recent update of the boot source.",
+			},
+			"url": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The URL of the boot source.",
+			},
+		},
+	}
+}
+
+func resourceBootSourceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client.Client)
+
+	bootsource, err := getBootSource(client)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId(fmt.Sprintf("%v", bootsource.ID))
+
+	// We create by changing the url to the chosen value
+	bootsourceParams := entity.BootSourceParams{
+		URL: d.Get("url").(string),
+	}
+
+	if _, err := client.BootSource.Update(bootsource.ID, &bootsourceParams); err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId(fmt.Sprintf("%v", bootsource.ID))
+
+	return resourceBootSourceRead(ctx, d, meta)
+}
+
+func resourceBootSourceRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client.Client)
+
+	bootsource, err := getBootSource(client)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(fmt.Sprintf("%v", bootsource.ID))
+
+	tfState := map[string]interface{}{
+		"created":          bootsource.Created,
+		"keyring_data":     bootsource.KeyringData,
+		"keyring_filename": bootsource.KeyringFilename,
+		"updated":          bootsource.Updated,
+		"url":              bootsource.URL,
+	}
+	if err := setTerraformState(d, tfState); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func resourceBootSourceUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client.Client)
+
+	bootsource, err := getBootSource(client)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId(fmt.Sprintf("%v", bootsource.ID))
+
+	bootsourceParams := entity.BootSourceParams{
+		KeyringData:     d.Get("keyring_data").(string),
+		KeyringFilename: d.Get("keyring_filename").(string),
+		URL:             d.Get("url").(string),
+	}
+
+	if _, err := client.BootSource.Update(bootsource.ID, &bootsourceParams); err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId(fmt.Sprintf("%v", bootsource.ID))
```

Comment:
This line is not needed. The Id is already set during the Create function's execution